### DISCUSSION TITLE
Tune Compaction and Backup

### DISCUSSION THREAD

**Discussion by ltagliamonte-dd** [apache/kvrocks#2379] [Q&A]
Hello folks! I'm in a pinch with kvrocks compaction/backup tuning....

I'm working on a data pipeline where I process a bunch of parquet files into kvrocks, every N files I create a checkpoint on S3 by calling compaction (COMPACT) on kvrocks and then backup (BGSAVE), and uploading the backup folder on S3.

Very often during the compaction and backup the memory consumption goes really high and my k8s pod with 40GB of mem get's oomkilled.
Is there a way to keep at bay the mem usage during compaction and backup? 
What are the knobs to tune? I don't see any knob for backups.

This is my first time using kvrocks and I'm not really familiar with all the config params.

**Comment by git-hulk** [comment]
@ltagliamonte-dd Could you provide rocksdb.* configurations in kvrocks.conf?

**Reply by ltagliamonte-dd** [reply]
Hello @git-hulk I'm using the cmd line options, this is my current config:
```
command=/usr/local/bin/kvrocks/kvrocks 
    --dir %(ENV_KVROCKS_DATA_DIR)s 
    --port 6379 
    --rocksdb.disable_auto_compactions yes 
    --rocksdb.rate_limiter_auto_tuned no 
    --rocksdb.max_background_jobs 12 
    --rocksdb.max_sub_compactions 6 
    --rocksdb.target_file_size_base 1024
```

**Reply by git-hulk** [reply]
@ltagliamonte-dd Thank you! Maybe you can try to decrease `max_background_jobs` and `max_sub_compactions` to avoid too many ongoing compaction threads. Guess you can just remove those two configurations in your command line.

**Reply by ltagliamonte-dd** [reply]
@git-hulk thank for the suggestion, my params are just 3X the defaults, i bumped those up because compaction on the dataset was taking >10minutes
is there any other params I can use to improve the compaction?
the backup also seems to consume a lot of memory, i wonder why kvrocks doesn't just symlink the data folder in the backup folder, like cassandra does for example.

**Reply by git-hulk** [reply]
> is there any other params I can use to improve the compaction?

Yes, this is a tough issue for rocksdb. It might consume too much memory and CPU while increasing the compaction threads, but decreasing it will affect the speed. But I'm not sure why do you have to do a compaction before creating the backup? not sure if you have many deleted/expired data inside the DB. I think it's a good idea to slow down the compaction speed to avoid consuming too many resources and wait until it's finished.

> the backup also seems to consume a lot of memory, i wonder why kvrocks doesn't just symlink the data folder in the backup folder, like cassandra does for example.

It was a hard symlink as well, so it should not consume much memory. Not sure if this's affected by compaction behavior.

**Reply by ltagliamonte-dd** [reply]
i do just inserts/updates, no deletes.
the problem with the compaction obviously is that they go over all the data, maybe I should just compact every N runs instead of doing it at every run.
is there a compaction strategy that would be "smarter" and touch just the files that got modified?

**Reply by git-hulk** [reply]
> is there a compaction strategy that would be "smarter" and touch just the files that got modified?

Kvrocks is now supported by the compaction checker and it will check if SST files need to be compacted periodically. Another point is that SST files are never updated and data are merged by compaction.

```
# The hour range that compaction checker would be active
# e.g. compaction-checker-range 0-7 means compaction checker would be worker between
# 0-7am every day.
compaction-checker-range 0-7
```

**Reply by ltagliamonte** [reply]
Thanks for the answers! Since I'm calling compaction at every run of my software I disabled them passing the flag to the cmd line.
My point about smarter compaction of the SST is that they all get reprocessed all the time isn't it?

**Reply by git-hulk** [reply]
> My point about smarter compaction of the SST is that they all get reprocessed all the time isn't it?

Oh, yes. That's right.

**Reply by git-hulk** [reply]
@ltagliamonte-dd Wanna know if everything is going well after reducing the number of compaction threads.

**Reply by ltagliamonte-dd** [reply]
Hello @git-hulk well yes and no, mem usage is contained however compaction take up to 60m to complete.
I'd like to find other params i can tune to reduce the memory footprint, like reduce the number in memory cache (i'm just writing in this kvrocks instance) and accelerate the compactions.

**Reply by git-hulk** [reply]
@ltagliamonte-dd The default value of memory cache is 4GiB and it only be used while having the read requests. So guess it would not help if the memory is consumed by the compaction. Another possible approach is to try to reduce `target_file_size_base` and increase a few compaction threads
